README.rst
=======
critics
=======

.. image:: https://img.shields.io/travis/coagulant/critics.svg
        :target: https://travis-ci.org/coagulant/critics

.. image:: https://img.shields.io/coveralls/coagulant/critics.svg
    :target: https://coveralls.io/r/coagulant/critics

.. image:: https://img.shields.io/pypi/v/critics.svg
        :target: https://pypi.python.org/pypi/critics

.. image:: https://img.shields.io/badge/licence-BSD-blue.svg

.. image:: https://img.shields.io/badge/status-beta-yellow.svg

::

    Usage: critics [OPTIONS]

      Notify about new reviews in AppStore and Google Play in slack.

      Launch command using supervisor or using screen/tmux/etc. Reviews are
      fetched for multiple apps and languages in --beat=300 interval.

    Options:
      --ios TEXT                   ios app id, e.g. 122434343
      --ios-channel TEXT           Slack channel for ios notifications, optional
      --android TEXT               Android app name, e.g. "com.rovio.angrybirds"
      --android-channel TEXT       Slack channel for android notifications, optional
      --language TEXT              ISO 639-1 languages of review [default: system locale]
      --slack-webhook TEXT         Slack webhook absolute URL, required
      --parse-max-entries INTEGER  Number of feed entries to look into [default: 10]
      --beat INTEGER               Number of seconds between polling feed [default: 300]
      --verbose / --short
      --notify / --no-notify
      --persist / --no-persist
      --model PATH
      --daemonize / --run-once
      --stats INTEGER              Port to serve prometheus stats [default: 9137]
      --version
      --help                       Show this message and exit.


Examples
~~~~~~~~

Track English and Russian reviews for iOS and Android Angry Birds apps
and post them to separate channels::

    critics --ios=343200656 --android=com.rovio.angrybirds \
            --slack-webhook=YOUR_SLACK_WEBHOOK_URL \
            --language=en --language=ru \
            --ios-channel="#ios_reviews" --android-channel="#android_reviews"

Previous command, but using env variables::

    CRITICS_IOS=343200656 CRITICS_ANDROID=com.rovio.angrybirds \
    CRITICS_SLACK_WEBHOOK=YOUR_SLACK_WEBHOOK_URL CRITICS_LANGUAGE="en ru" \
    CRITICS_IOS_CHANNEL="#ios_reviews" CRITICS_ANDROID_CHANNEL="#android_reviews" \
    critics

For demo purpose (does not require slack): parse ios feed for MyBook app
and print reviews in stdout::

    critics --ios=556540446 --run-once --no-notify --no-persist --verbose

Monitor several ios apps to distinct slack channels::

    critics --ios=343200656 --ios=343200657 \
            --slack-webhook=YOUR_SLACK_WEBHOOK_URL \
            --ios-channel="#ios_reviews_appA" --ios-channel="#ios_reviews_appB"

How it looks
~~~~~~~~~~~~

.. image:: https://github.com/coagulant/critics/raw/master/docs/screenshots/screenshot_01.png
   :height: 356 px
   :width: 492 px
   :alt: Heroes HD Android reviews


CHANGELOG.rst
Changelog
---------

1.0.0 (-------TBA)
------------------
* Fix crash when app has no android reviews
* Add sentry support via --sentry-dsn

0.3.1 (2017-01-26)
------------------
* Fix test build and readme
* Add docker support


0.3.0 (2016-03-31)
------------------
* Fix crashes when no locale env variables were set
* Add prometheus metrics
* Multiple channels for multiple apps on same platform


0.2.1 (2015-03-20)
------------------
* Change default request timeout to 5 seconds

0.2.0 (2015-03-19)
------------------

* Multiple languages support
* English and Russian translations
* Usable url for iOS review


0.1.0 (2015-03-09)
------------------

* Initial release


tests/test_cli.py
# coding: utf-8
import json

import sys
from click.testing import CliRunner
import pytest

from critics import envvar_prefix
from critics.commands import cli


def test_no_locale():
    runner = CliRunner()
    result = runner.invoke(cli, ['--run-once'], env={'LC_CTYPE': '', 'LC_ALL': '', 'LANG': '', 'LANGUAGE': ''})
    exit_code = -1 if sys.version_info >= (3, 0) else 0  # click aborts if locale is not set up
    assert result.exit_code == exit_code


@pytest.mark.functional
def test_args(tmpdir):
    model = tmpdir.join("reviews.json")
    runner = CliRunner()
    result = runner.invoke(cli, ['--ios', '343200656',
                                 '--android', 'com.rovio.angrybirds',
                                 '--language', 'ru',
                                 '--model', model.strpath,
                                 '--run-once'],
                           env={'CRITICS_SLACK_WEBHOOK': 'http://httpbin.org/post',
                                'CRITICS_IOS_CHANNEL': '#ios',
                                'CRITICS_ANDROID_CHANNEL': '#android'},
                           auto_envvar_prefix=envvar_prefix)
    assert result.exit_code == 0

    # iOS rss feed returns either 10 or 0 reviews - probably a bug on Apple side,
    # sometimes the feed is just empty.
    num_reviews_ios = len(json.loads(model.read())['ios']['value'])
    num_reviews_android = len(json.loads(model.read())['android']['value'])

    assert num_reviews_ios in [0, 10]
    assert num_reviews_android == 10
    assert result.output == ('Languages: Russian\n'
                             'Tracking IOS apps: 343200656\n'
                             'Tracking Android apps: com.rovio.angrybirds\n'
                             'Transport: slack channels:   343200656 -> #ios   com.rovio.angrybirds -> #android\n'
                             'ios: 343200656: ru: Fetched %s reviews, %s new\n' % (num_reviews_ios, num_reviews_ios) +
                             'android: com.rovio.angrybirds: ru: Fetched 10 reviews, 10 new\n')


@pytest.mark.functional
def test_mutiple_languages(tmpdir):
    model = tmpdir.join("reviews.json")
    runner = CliRunner()
    result = runner.invoke(cli, ['--android', 'com.rovio.angrybirds',
                                 '--android', 'com.rovio.angrybirdsstarwarsii.ads',
                                 '--model', model.strpath,
                                 '--run-once', '--no-notify'],
                           env={'CRITICS_LANGUAGE': 'ru en'},
                           auto_envvar_prefix=envvar_prefix)
    assert result.exit_code == 0
    assert result.output == ('Languages: Russian, English\n'
                             'Tracking Android apps: com.rovio.angrybirds, com.rovio.angrybirdsstarwarsii.ads\n'
                             'android: com.rovio.angrybirds: ru: Fetched 10 reviews, 10 new\n'
                             'android: com.rovio.angrybirds: en: Fetched 10 reviews, 10 new\n'
                             'android: com.rovio.angrybirdsstarwarsii.ads: ru: Fetched 10 reviews, 10 new\n'
                             'android: com.rovio.angrybirdsstarwarsii.ads: en: Fetched 10 reviews, 10 new\n')


tests/test_core.py
# coding: utf-8
import json
try:
    from unittest.mock import Mock
except Exception:
    from mock import Mock
from critics.core import CriticApp
from critics.parsers import Review


def test_poll_store(tmpdir):
    fakemodel = tmpdir.join("fakemodel.json")
    app = CriticApp(ios=['app1', 'app2'], language=['ru'], persist=True, slack_webhook='http://www')
    app.model_file = fakemodel.strpath
    fake_fetcher = Mock(return_value=[Review(
        id=u'xfkew4ytwqqddid:2e22wdad',
        platform='ios',
        title=u'Great app! ♡',
        rating=2,
        summary=u'Here be\nDragons!',
        url=u'http://www',
        author=u'Here comes more BS',
        date='bull',
        language='en',
        version=None
    )])
    app.fetchers['ios'] = fake_fetcher
    fake_notifier = Mock()
    app.notifiers['slack'] = fake_notifier
    app.poll_store(platform='ios')

    assert fake_fetcher.call_count == 2
    assert fake_notifier.call_count == 1
    assert len(app.reviews['ios']) == 1

    fake_fetcher.reset_mock()
    fake_notifier.reset_mock()
    app.poll_store(platform='ios')
    assert fake_fetcher.call_count == 2
    assert fake_notifier.call_count == 0
    assert len(app.reviews['ios']) == 1


def test_load_model():
    app = CriticApp(persist=False)
    assert not app.load_model()
    assert app.reviews['ios'] == set()
    assert app.reviews['android'] == set()

    app = CriticApp(persist=True, model='tests/fixtures/model.json')
    assert app.load_model()
    assert app.reviews['ios'] == {'123'}
    assert app.reviews['android'] == {'xxx', 'yyyy'}


def test_save_model(tmpdir):
    fakemodel = tmpdir.join("fakemodel.json")
    app = CriticApp(persist=False)

    assert not app.save_model()
    assert not fakemodel.check()

    app = CriticApp(persist=True, model=fakemodel.strpath)

    assert app.save_model()
    assert fakemodel.check()
    assert json.load(fakemodel) == {'android': {'python_object': 'set', 'value': []},
                                    'ios': {'python_object': 'set', 'value': []}}

    app.reviews['ios'].add('678')
    app.reviews['android'].add('qqq')
    assert app.save_model()
    assert fakemodel.check()
    assert json.load(fakemodel) == {'android': {'python_object': 'set', 'value': ['qqq']},
                                    'ios': {'python_object': 'set', 'value': ['678']}}


tests/test_parsers.py
# coding: utf-8
import codecs
import datetime
import responses
from critics.parsers import get_ios_reviews, get_android_reviews


@responses.activate
def test_ios():
    responses.add(responses.GET, 'https://itunes.apple.com/fr/rss/customerreviews/id=123/sortBy=mostRecent/xml',
                  body=codecs.open('tests/fixtures/itunes_fr.example', encoding='utf-8').read(),
                  content_type='application/xml; charset=UTF-8')
    reviews = get_ios_reviews(123, language='fr')

    assert len(reviews) == 50

    review = reviews[0]
    assert review.id == u'1163882289'
    assert review.title == u'Leon, super jeu'
    assert review.rating == 5
    assert review.summary == u'Moi qui ne suis pas n grand amateur, j\'ai adoré ce jeu malin et intuitif.'
    assert review.url == u'https://itunes.apple.com/fr/reviews/id270119737'
    assert review.author == u'Justine M18'
    assert review.date == datetime.datetime(2015, 3, 5, 20, 5)
    assert review.version == '1.7'

    assert str(review) == """Review (1163882289):
title=Leon, super jeu
rating=5
summary=Moi qui ne suis pas n grand amateur, j'ai adoré ce jeu malin et intuitif.
url=https://itunes.apple.com/fr/reviews/id270119737
author=Justine M18
date=2015-03-05 20:05:00
language=fr
version=1.7"""


@responses.activate
def test_android_ru():
    responses.add(responses.POST, 'https://play.google.com/store/getreviews',
                  body=codecs.open('tests/fixtures/gp_ru.example', encoding='utf-8').read(),
                  content_type='application/json; charset=UTF-8')
    reviews = get_android_reviews('com.skype.raider', language=None, limit=10)

    assert 'hl=en' not in responses.calls[0].request.body
    assert len(reviews) == 10

    review = reviews[0]
    assert review.id == u'gp:AOqpTOGVTpUWQyZJhCbF6dpGfgWnmaj992am2upoCj8Iur1YoNKubAlN_W_BzRYHGh7pdxAe-HDIdOUEvaRfpw'
    assert review.title == u'Не работает'
    assert review.rating == 1
    assert review.summary == u'Не работает видео связь на nvidia shield'
    assert review.url == u'https://play.google.com/store/apps/details?id=com.skype.raider&reviewId=Z3A6QU9xcFRPR1ZUcFVXUXlaSmhDYkY2ZHBHZmdXbm1hajk5MmFtMnVwb0NqOEl1cjFZb05LdWJBbE5fV19CelJZSEdoN3BkeEFlLUhESWRPVUV2YVJmcHc'  # noqa
    assert review.author == u'Polar Alexander'
    assert review.date == u'7 марта 2015 г.'
    assert review.version is None

    assert reviews[1].rating == 2


@responses.activate
def test_android_en():
    responses.add(responses.POST, 'https://play.google.com/store/getreviews',
                  body=codecs.open('tests/fixtures/gp_en.example', encoding='utf-8').read(),
                  content_type='application/json; charset=UTF-8')
    reviews = get_android_reviews('com.skype.raider', language='en', limit=10)

    assert 'hl=en' in responses.calls[0].request.body
    assert len(reviews) == 10

    review = reviews[0]
    assert review.id == u'gp:AOqpTOFuJk3lGt-xaYQLrp03KTe8eg7qs7Bm9wkUZQXQ1b2knLXIdpq1e9ZPM_AwCEa3szAZH9DDqWPbVE5qBQ'
    assert review.title == u''
    assert review.summary == u'I like skype'
    assert review.date == u'March 14, 2015'
    assert review.version is None


@responses.activate
def test_android_empty():
    responses.add(responses.POST, 'https://play.google.com/store/getreviews',
                  body=codecs.open('tests/fixtures/gp_empty.example', encoding='utf-8').read(),
                  content_type='application/json; charset=UTF-8')
    reviews = get_android_reviews('ru.zvukislov', language='en')
    assert len(reviews) == 0


critics/core.py
import json
import logging
import os

import tornado.web
from prometheus_client import Summary, Counter, Gauge

from .parsers import get_android_reviews, get_ios_reviews
from .transport import post2slack

logger = logging.getLogger('critics')


fetches = Summary('critics_fetches_seconds', 'Time spent, fetching data', ['platform'])
notifications_counter = Counter('critics_notifications_total', 'How many chat messages sent', ['platform'])
reviews_counter = Counter('critics_reviews_total', 'Number of new reviews fetched', ['platform'])
last_scrape = Gauge('critics_last_run', 'Last time successful scrape')
last_review = Gauge('critics_last_review', 'Last time new review fetched')


class CriticApp(tornado.web.Application):

    def __init__(self, *args, **kwargs):
        super(CriticApp, self).__init__(*args, **kwargs)
        self.reviews = {
            'android': set(),
            'ios': set(),
        }
        self.fetchers = {
            'android': get_android_reviews,
            'ios': get_ios_reviews
        }
        self.notifiers = {
            'slack': post2slack
        }
        self.channels = kwargs.get('channels', {})

    def poll_store(self, platform, notify=True):
        for app_id in self.settings.get(platform):
            for language in self.settings['language']:
                try:
                    self.poll_store_single_app(platform, app_id, language, notify)
                except Exception:
                    if self.sentry_client:
                        self.sentry_client.captureException()
                    raise
        last_scrape.set_to_current_time()

    def poll_store_single_app(self, platform, app_id, language, notify):
        if platform not in self.fetchers.keys():
            raise Exception('Unknown platform %s', platform)
        fetcher = self.fetchers[platform]

        new_reviews = []
        parsed_review_ids = self.reviews[platform]
        logging.debug('parsed_review_ids = %s', parsed_review_ids)
        with fetches.labels({'platform': platform}).time():
            reviews = fetcher(app_id=app_id,
                              language=language,
                              limit=self.settings.get('parse_max_entries', None))

        for review in reviews:
            if review.id in parsed_review_ids:
                continue
            logger.debug(review)
            parsed_review_ids.add(review.id)
            new_reviews.append(review)

        logger.info('%(platform)s: %(app_id)s: %(language)s: Fetched %(num_reviews)s reviews, %(new_reviews)s new',
                    {'platform': platform, 'app_id': app_id, 'language': language,
                     'num_reviews': len(reviews), 'new_reviews': len(new_reviews)})

        if new_reviews:
            last_review.set_to_current_time()
            self.send_messages(new_reviews, platform, notify, app_id)

    def get_channel(self, platform, app_id):
        platform_channels = self.channels.get(platform, {})
        return platform_channels.get(app_id, None)

    def send_messages(self, new_reviews, platform, notify, app_id):
        self.save_model()
        channel = self.get_channel(platform, app_id)
        if not notify:
            logger.debug('%s: Skip notification on first run', platform)
        elif not self.settings.get('notify', True):
            logger.debug('%(platform)s: Faking %(num)s notifications to %(channel)s', {
                'platform': platform, 'num': len(new_reviews), 'channel': channel
            })
        else:
            logging.debug(new_reviews)
            notifications_counter.labels({'platform': platform}).inc()
            reviews_counter.labels({'platform': platform}).inc(len(new_reviews))
            self.notifiers['slack'](new_reviews, self.settings['slack_webhook'], channel)

    def load_model(self):
        if not self.settings['persist']:
            return False
        if os.path.exists(self.settings['model']):
            logger.debug('Loading state')
            try:
                with open(self.settings['model']) as fp:
                    self.reviews = json.load(fp, object_hook=as_set)
            except Exception as e:
                logger.error('Problem loading prev. results from file: %s', e)
            else:
                return True
        return False

    def save_model(self):
        if not self.settings['persist']:
            return False
        logger.debug('Saving state')
        try:
            json.dump(self.reviews, open(self.settings['model'], 'w'), cls=SetEncoder, indent=4)
        except Exception as e:
            logger.error('Problem saving results to file: %s', e)
            return False
        else:
            return True


class SetEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, set):
            return {'python_object': 'set', 'value': list(obj)}
        return json.JSONEncoder.default(self, obj)


def as_set(dct):
    if 'python_object' in dct:
        return set(dct['value'])
    return dct


critics/parsers.py
# coding: utf-8
from collections import namedtuple
import json
import datetime
import logging
import os
import re
from time import mktime

import feedparser
import requests
from lxml import html, etree

from .compat import python_2_unicode_compatible


logger = logging.getLogger('critics')
timeout = os.environ.get('CRITICS_TIMEOUT', 5)
utf8_parser = html.HTMLParser(encoding='utf-8')


@python_2_unicode_compatible
class Review(namedtuple('Review',
                        ['id', 'platform', 'title', 'rating', 'summary', 'url',
                         'author', 'date', 'language', 'version'])):
    __slots__ = ()

    def __str__(self):
        return (u'Review (%s):\ntitle=%s\nrating=%s\nsummary=%s\nurl=%s\n'
                u'author=%s\ndate=%s\nlanguage=%s\nversion=%s' % (
                    self.id,
                    self.title,
                    self.rating,
                    self.summary,
                    self.url,
                    self.author,
                    self.date,
                    self.language,
                    self.version
                ))


def get_ios_reviews(app_id, language, limit=100):
    url = 'https://itunes.apple.com/%(language)srss/customerreviews/id=%(app_id)s/sortBy=mostRecent/xml' % {
        'language': '%s/' % language if language else '', 'app_id': app_id}
    response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1)'},
                            timeout=timeout)
    response.encoding = 'utf-8'  # avoid chardet not guessing correctly
    feed = feedparser.parse(response.text)
    reviews = [Review(
        id=entry.id,
        platform='ios',
        title=entry.title,
        rating=int(entry.im_rating),
        summary=entry.summary,
        url=entry.href,
        author=entry.author,  # author url: entry.href
        date=datetime.datetime.fromtimestamp(mktime(entry.updated_parsed)),
        language=language,
        version=entry.im_version
    ) for entry in feed['entries'][1:1 + limit]]
    return reviews


def get_android_reviews(app_id, language, limit=100):
    url = 'https://play.google.com/store/getreviews'
    payload = {'xhr': 1, 'id': app_id, 'reviewSortOrder': 0, 'pageNum': 0, 'reviewType': 0}
    if language:
        payload['hl'] = language
    response = requests.post(url, data=payload, timeout=timeout)
    json_source = response.text[response.text.find('['):]
    response_as_json = json.loads(json_source)
    try:
        response_as_html = response_as_json[0][2]
    except IndexError:
        logger.error('Unexpected json for app_id=%s', app_id, exc_info=True)
        return []

    # no reviews
    if not response_as_html:
        return []

    try:
        doc = html.fromstring(response_as_html.encode('utf-8'), parser=utf8_parser)
        reviews_html = doc.cssselect('.single-review')
    except etree.XMLSyntaxError:
        logger.error('Unparsable html', exc_info=True)
        return []

    def get_rating_from_html(review_html):
        star_style = review_html.cssselect('.current-rating')[0].get('style')  # e.g. 'width: 20%'
        return int(re.search('(\d+)%', star_style).group(1)) / 20

    reviews = [Review(
        id=review_html.cssselect('.review-header')[0].get('data-reviewid'),
        platform='android',
        title=review_html.cssselect('.review-body .review-title')[0].text_content().strip(),
        rating=get_rating_from_html(review_html),
        summary=review_html.cssselect('.review-body .review-title')[0].tail.strip(),
        url='https://play.google.com' + review_html.cssselect('.reviews-permalink')[0].get('href'),
        author=review_html.cssselect('.review-header .author-name')[0].text_content().strip(),
        date=review_html.cssselect('.review-header .review-date')[0].text_content().strip(),
        language=language,
        version=None
    ) for review_html in reviews_html[:limit]]
    return reviews


critics/transport.py
# coding: utf-8
import json
import datetime
import math

from babel.dates import format_datetime
import requests

from .i18n import get_language, get_locale


locale = get_locale()
language = get_language()
_ = language.ugettext


def post2slack(reviews, slack_url, channel):
    emoji = {
        1: ':feelsgood:',
        2: ':finnadie:',
        3: ':hurtrealbad:',
        4: ':suspect:',
        5: ':godmode:',
    }
    colors = {
        1: '#CC2525',
        2: '#CC2525',
        3: '#E8AD0C',
        4: '#30E80C',
        5: '#30E80C',
    }
    stars = {
        1: u'★☆☆☆☆',
        2: u'★★☆☆☆',
        3: u'★★★☆☆',
        4: u'★★★★☆',
        5: u'★★★★★'
    }

    if not reviews:
        return

    def get_date_string(date):
        if isinstance(date, datetime.datetime):
            return format_datetime(date, 'd MMMM yyyy hh:mm', locale=get_locale())
        else:
            return date

    average_rating = int(math.floor(sum(review.rating for review in reviews) / float(len(reviews))))
    platform = reviews[0].platform

    store = _('AppStore') if platform == 'ios' else _('Google Play')
    template = language.ungettext('There is %(num)d new review in %(store)s',
                                  'There are %(num)d new reviews in %(store)s', len(reviews))
    text = template % {'num': len(reviews), 'store': store}

    payload = {
        'attachments': [{
            'color': colors[review.rating],
            'title': review.title or u'-',
            'title_link': review.url,
            'text': u'{stars}\n{summary}\n\n_{author}_, {date} {version}'.format(
                stars=stars[review.rating],
                summary=review.summary,
                author=review.author,
                date=get_date_string(review.date),
                version=u'[%s]' % review.version if review.version else ''
            ),
            "mrkdwn_in": ["text"],
        } for review in reviews],
        'text': text,
        'username': _('Critic'),
        "icon_emoji": emoji[average_rating],
    }
    if channel:
        payload['channel'] = channel

    requests.post(slack_url, data={'payload': json.dumps(payload)})


critics/commands.py
#!/usr/bin/env python
# coding: utf-8
import logging
from collections import OrderedDict
from functools import partial
from itertools import chain, repeat

import click
import tornado.httpserver
import tornado.ioloop
import tornado.web
from babel import Locale, UnknownLocaleError
from prometheus_client import start_http_server
from raven import Client

import critics
from .core import CriticApp
from .i18n import get_locale

logger = logging.getLogger('critics')


@click.command()
@click.option('--ios', multiple=True, help='ios app id, e.g. 122434343')
@click.option('--ios-channel', multiple=True, help='Slack channel for ios notifications, optional')
@click.option('--android', multiple=True, help='Android app name, e.g. "com.rovio.angrybirds"')
@click.option('--android-channel', multiple=True, help='Slack channel for android notifications, optional')
@click.option('--language', multiple=True, help='ISO 639-1 languages of review [default: system locale]')
@click.option('--slack-webhook', help='Slack webhook absolute URL, required')
@click.option('--parse-max-entries', default=10, help='Number of feed entries to look into [default: 10]')
@click.option('--beat', default=300, help='Number of seconds between polling feed [default: 300]')
@click.option('--verbose/--short', default=False)
@click.option('--notify/--no-notify', default=True)
@click.option('--persist/--no-persist', default=True)
@click.option('--model', type=click.Path(), default='reviews.json')
@click.option('--daemonize/--run-once', default=True)
@click.option('--stats', default=9137, help='Port to serve prometheus stats [default: 9137]')
@click.option('--sentry-dsn', help='DSN of Sentry instance to monitor exceptions')
@click.option('--version', is_flag=True)
def cli(**settings):
    """Notify about new reviews in AppStore and Google Play in slack.

       Launch command using supervisor or using screen/tmux/etc.
       Reviews are fetched for multiple apps and languages in --beat=300 interval.
    """
    setup_logging(settings)
    settings = setup_languages(settings)
    channels = setup_channel_map(settings)
    app = CriticApp(**dict(settings, channels=channels))
    if settings['sentry_dsn']:
        app.sentry_client = Client(settings['sentry_dsn'])
        logger.debug('Errors are reported to %s' % settings['sentry_dsn'])
    else:
        app.sentry_client = None

    if settings['version']:
        click.echo('Version %s' % critics.__version__)
        return
    if not (settings['ios'] or settings['android']):
        click.echo('Please choose either --ios or --android')
        return

    loop = tornado.ioloop.IOLoop.instance()

    if app.load_model():
        logger.debug('Model loaded OK, not skipping notify on first run')
        notify = True
    else:
        notify = False

    if settings['ios']:
        logger.info('Tracking IOS apps: %s', ', '.join(settings['ios']))
        itunes = tornado.ioloop.PeriodicCallback(partial(app.poll_store, 'ios'),
                                                 1000 * settings['beat'], loop)
        itunes.start()
    if settings['android']:
        logger.info('Tracking Android apps: %s', ', '.join(settings['android']))
        google_play = tornado.ioloop.PeriodicCallback(partial(app.poll_store, 'android'),
                                                      1000 * settings['beat'], loop)
        google_play.start()

    echo_channel_map(channels)

    if settings['ios']:
        app.poll_store('ios', notify=notify)
    if settings['android']:
        app.poll_store('android', notify=notify)

    if settings['stats']:
        port = int(settings['stats'])
        logger.debug('Serving metrics server on port %s' % port)
        start_http_server(port)

    if settings['daemonize']:
        loop.start()


def setup_logging(settings):
    handler = logging.StreamHandler()
    logger.addHandler(handler)
    if settings['verbose']:
        logger.setLevel(logging.DEBUG)
        handler.setFormatter(logging.Formatter('[%(asctime)s] %(message)s', "%Y-%m-%d %H:%M:%S"))
    else:
        logger.setLevel(logging.INFO)
    logger.propagate = False


def setup_languages(settings):
    if not settings['language']:
        settings['language'] = [get_locale()[:2]]

    languages = []
    language_names = []
    for lang_code in settings['language']:
        try:
            language_names.append(Locale(lang_code).english_name)
            languages.append(lang_code)
        except UnknownLocaleError:
            raise click.ClickException('Unknown language code: %s' % lang_code)

    logger.info('Languages: %s', ', '.join(language_names))
    return settings


def setup_channel_map(settings):
    channel_map = OrderedDict()
    if not settings['slack_webhook']:
        return channel_map
    if settings['ios'] and settings['ios_channel']:
        channel_map['ios'] = {}
        ios_channels = chain(settings['ios_channel'], repeat(settings['ios_channel'][-1]))
        for app_id, channel in zip(settings['ios'], ios_channels):
            channel_map['ios'][app_id] = channel
    if settings['android'] and settings['android_channel']:
        channel_map['android'] = {}
        android_channels = chain(settings['android_channel'], repeat(settings['android_channel'][-1]))
        for app_id, channel in zip(settings['android'], android_channels):
            channel_map['android'][app_id] = channel
    return channel_map


def echo_channel_map(channel_map):
    if not channel_map:
        return
    channel_output = ''
    for platform, app_mapping in channel_map.items():
        for app_id, channel in app_mapping.items():
            channel_output += '   {app_id} -> {channel}'.format(
                platform=platform, app_id=app_id, channel=channel)
    logger.info('Transport: slack channels:%s' % channel_output)
